transcendence-app/src/shared/db/postgres/postgresConnection.provider.ts

```diff
@@ -45,4 +46,27 @@ export class PostgresPool implements OnModuleInit, OnModuleDestroy {
   query(query: Query): Promise<QueryResult<any>> {
     return this.pool.query(query);
   }
+
+  async transaction<T>(callback: (client: PoolClient) => Promise<T | null>) {
+    let data: T | null = null;
+    try {
+      const client = await this.pool.connect();
+      try {
+        await client.query('BEGIN');
+        data = await callback(client);
+        await client.query('COMMIT');
+      } catch (err) {
+        await client.query('ROLLBACK');
+        throw err;
+      } finally {
+        client.release();
+      }
+      return data;
+    } catch (err) {
+      if (isCriticalDatabaseError(err)) {
+        this.logger.error(err.message);
```

Comment:
I don't get why do we only log 'FATAL' or 'PANIC' severity level errors.

Reply:
https://www.postgresql.org/docs/current/runtime-config-logging.html#RUNTIME-CONFIG-SEVERITY-LEVELS

The 'pg' driver throws errors in many cases. For example, when the query has a syntax error or failed because of some constrain.

Let's take an example of a user who wants to update his email address and select an email address already in use. On the Users database table, the email column has a UNIQUE constrain. In this case, the 'pg' driver throws an error.

If we add logging to every case, it could get noisy and hard to notice critical errors. Also, the database container already logs these cases.

Here we log unrecoverable errors, like if the database connection is down or it can't save data because there is no space left. Anyway, this is my opinion. If we decide that it's an improvement to log every error, we could remove this condition. Or maybe log all the errors in the development environment and leave this condition for the production environment.